iceoryx_utils/test/moduletests/test_cxx_deadline_timer.cpp

```diff
@@ -0,0 +1,108 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_utils/cxx/deadline_timer.hpp"
+#include "iceoryx_utils/internal/units/duration.hpp"
+#include "test.hpp"
+#include "testutils/timing_test.hpp"
+
+#include <chrono>
+#include <thread>
+
+using namespace ::testing;
+
+using namespace iox::units;
+using namespace iox::units::duration_literals;
+
+using Timer = iox::cxx::DeadlineTimer;
+
+class DeadlineTimer_test : public Test
+{
+  public:
+    virtual void SetUp()
+    {
+        numberOfCalls = 0;
+    }
+
+    virtual void TearDown()
+    {
+    }
+
+    Duration second{1_s};
+
+    std::atomic<int> numberOfCalls{0};
+    static const Duration TIMEOUT;
+    static const int SLEEPTIME;
+};
+
+const Duration DeadlineTimer_test::TIMEOUT{10_ms};
+const int DeadlineTimer_test::SLEEPTIME = DeadlineTimer_test::TIMEOUT.milliSeconds<int>();
+
+TIMING_TEST_F(DeadlineTimer_test, DurationOfNonZeroIsExpiresAfterTimeout, Repeat(5), [&] {
+    Timer sut(TIMEOUT);
+
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME / 3));
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME / 3));
+    TIMING_TEST_EXPECT_TRUE(sut.hasExpired());
+});
+
+TIMING_TEST_F(DeadlineTimer_test, ResetWithDurationIsExpired, Repeat(5), [&] {
+    Timer sut(TIMEOUT);
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME));
+    TIMING_TEST_EXPECT_TRUE(sut.hasExpired());
+    sut.reset();
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+});
+
+TIMING_TEST_F(DeadlineTimer_test, ResetWhenNotExpiredIsStillNotExpired, Repeat(5), [&] {
+    Timer sut(TIMEOUT);
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME / 3));
+    sut.reset();
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME / 3));
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+});
+
+TIMING_TEST_F(DeadlineTimer_test, ResetAfterBeingExpiredIsNotExpired, Repeat(5), [&] {
+    Timer sut(TIMEOUT);
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME));
+
+    TIMING_TEST_ASSERT_TRUE(sut.hasExpired());
+    sut.reset();
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+});
+
+TIMING_TEST_F(DeadlineTimer_test, ResetWithCustomizedTimeAfterBeingExpiredIsNotExpired, Repeat(5), [&] {
+    Timer sut(TIMEOUT);
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME));
+
+    TIMING_TEST_ASSERT_TRUE(sut.hasExpired());
+
+    sut.reset(20_s);
+    std::this_thread::sleep_for(std::chrono::milliseconds(2 * SLEEPTIME));
+
+    TIMING_TEST_EXPECT_FALSE(sut.hasExpired());
+});
+
+
+TIMING_TEST_F(DeadlineTimer_test, RemainingTimeCheckIfNotExpired, Repeat(5), [&] {
```

Comment:
Could you please write a test where you verify that the remaining time is zero when the timer is expired.

Reply:
Okay